prog/minimization.go

```diff
@@ -121,11 +121,21 @@ func (typ *UnionType) minimize(ctx *minimizeArgsCtx, arg Arg, path string) bool
 }
 
 func (typ *PtrType) minimize(ctx *minimizeArgsCtx, arg Arg, path string) bool {
-	// TODO: try to remove optional ptrs
 	a := arg.(*PointerArg)
 	if a.Res == nil {
 		return false
 	}
+	if !ctx.triedPaths[path] {
```

Comment:
Is this check necessary? I would expect minimizeArgsCtx.do to take care of this already.

Reply:
I think we need a different `path` here, like `path + "->"` for example. The idea is that we try to minimize the pointer and if we fail, we remember that we tried by setting `ctx.triedPaths[path + "->"]` and `return true` to retry since we modified `ctx.p`.

Reply:
This is the idea behind current `triedPaths[path]` handling. Why do we need `path + "->"` on top of that?
You are minimizing the pointer arg. We already have general mechanism to remember if we tried to minimize any arg or not. Just use it.

Reply:
When we try to minimize a pointer we modify `ctx.p`, so we need to `return true` to retry whether minimization failed or succeeded. If minimization failed we need to somehow remember that we already tried it, hence `triedPaths[path + "->"] = true`. Otherwise we would just try it again, as `ctx.do` sets `ctx.triedPaths[path] = true` only when `minimize` returns `false`. We do the same thing for `ArrayType`.

Reply:
You are right.